src/proxy/inbound.rs

```diff
@@ -98,13 +103,21 @@ impl Inbound {
                 .http2_initial_stream_window_size(self.cfg.window_size)
                 .http2_initial_connection_window_size(self.cfg.connection_window_size)
                 .http2_max_frame_size(self.cfg.frame_size)
-                .serve(service);
+                .serve(service)
+                .with_graceful_shutdown(async {
+                    let shutdown = self.drain.signaled().await;
+                    // TODO: add some more graceful shutdown way
+                    time::sleep(time::Duration::from_secs(3)).await;
+                    info!("drain end");
+                    drop(shutdown);
+                });
 
             info!("HBONE listener established {}", addr);
 
             if let Err(e) = server.await {
                 error!("server error: {}", e);
             }
+            info!("server exit success drained");
```

Comment:
This still doesn't ever get reached. We need to move `drop(shutdown)` until after the `server.await` or we aren't actually waiting for the shutdown (aside from the arbitrary 3s sleep)

Reply:
>We need to move drop(shutdown) until after the server.await
I think it make sense